src/olm/account/mod.rs

```diff
@@ -439,6 +453,102 @@ impl Account {
+    ///
+    /// A dehydrated device is a device that is stored encrypted on the server
+    /// that can receive messages when the user has no other active devices.
+    /// Upon login, the user can rehydrate the device (using
+    /// [`from_dehydrated_device`]) and decrypt the messages sent to the
+    /// dehydrated device.
+    ///
+    /// The account must be a newly-created account that does not have any Olm
+    /// sessions, since the dehydrated device format does not store sessions.
+    ///
+    /// Returns the ciphertext and nonce.  `key` is a 256-bit (32-byte) key for
+    /// encrypting the device.
+    ///
+    /// The format used here is defined in
+    /// [MSC3814](https://github.com/matrix-org/matrix-spec-proposals/pull/3814).
+    pub fn to_dehydrated_device(
+        &self,
+        key: &[u8; 32],
+    ) -> Result<DehydratedDeviceResult, crate::DehydratedDeviceError> {
+        use matrix_pickle::Encode;
+
+        use self::dehydrated_device::Pickle;
+        use crate::utilities::base64_encode;
+
+        let pickle: Pickle = self.into();
+        let mut encoded = pickle.encode_to_vec()?;
+        let cipher = ChaCha20Poly1305::new(key.into());
+        let rng = thread_rng();
+        let nonce = ChaCha20Poly1305::generate_nonce(rng);
+        let ciphertext = cipher.encrypt(&nonce, encoded.as_slice());
+        encoded.zeroize();
+        let ciphertext = ciphertext?;
+
+        Ok(DehydratedDeviceResult {
+            ciphertext: base64_encode(ciphertext),
+            nonce: base64_encode(nonce),
+        })
+    }
+
+    /// Create an [`Account`] object from a dehydrated device.
+    ///
+    /// `ciphertext` and `nonce` are the ciphertext and nonce returned by
+    /// [`to_dehydrated_device`]. `key` is a 256-bit (32-byte) key for
+    /// decrypting the device, and must be the same key used when
+    /// [`to_dehydrate_device`] was called.
+    pub fn from_dehydrated_device(
+        ciphertext: &str,
+        nonce: &str,
+        key: &[u8; 32],
+    ) -> Result<Self, crate::DehydratedDeviceError> {
+        use std::io::Cursor;
+
+        use matrix_pickle::Decode;
+
+        use self::dehydrated_device::{Pickle, PICKLE_VERSION};
+        use crate::utilities::{base64_decode, get_pickle_version};
+
+        let cipher = ChaCha20Poly1305::new(key.into());
+        let ciphertext = base64_decode(ciphertext)?;
+        let nonce = base64_decode(nonce)?;
+        if nonce.len() != 12 {
+            return Err(crate::DehydratedDeviceError::InvalidNonce);
+        }
+        let mut plaintext = cipher.decrypt(nonce.as_slice().into(), ciphertext.as_slice())?;
+        let version =
+            get_pickle_version(&plaintext).ok_or(crate::DehydratedDeviceError::MissingVersion)?;
+        if version != PICKLE_VERSION {
+            return Err(crate::DehydratedDeviceError::Version(PICKLE_VERSION, version));
+        }
+
+        let mut cursor = Cursor::new(&plaintext);
```

Comment:
Can't the `from_decrypted_dehydrated_device()` function be used here? Otherwise we're fuzzing a function that we're not actually using.

Reply:
This has been implemented.